internal/report/report.go

```diff
@@ -179,7 +214,106 @@ func (r *ClusterPolicyReport) AddResult( //nolint:dupl
 	).Msg("added result to report")
 }
 
-func newPolicyReportResult( //nolint:funlen,cyclop
+func (r *ClusterPolicyReport) CopyResult(result *v1alpha2.PolicyReportResult) {
+	switch result.Result {
+	case StatusFail:
+		r.Summary.Fail++
+	case StatusError:
+		r.Summary.Error++
+	case StatusPass:
+		r.Summary.Pass++
+	}
+	r.Results = append(r.Results, result)
+}
+
+// GetReusablePolicyReportResult tries to find a PolicyReportResult that
+// can be reused.
+//
+// The result can be reused if both these conditions are
+// satisfied:
+//   - The subject of the PolicyReportResult (the object that was inspected)
+//     has not been changed since the report was created
+//   - The policy that evaluated the subject (now given by the user as
+//     parameter) has not been changed since the report was created
+func (r *ClusterPolicyReport) GetReusablePolicyReportResult(policy policiesv1.Policy, resource unstructured.Unstructured) *v1alpha2.PolicyReportResult {
+	return findReusableResult(r.Results, policy, resource)
+}
+
+// isReportGeneratedByPolicy checks if the given PolicyReportResult
+// has been generated by the given policy.
+// The comparison uses the policy UID and its revision and checks them
+// with the metadata stored inside of the given PolicyReportResult
+func isReportGeneratedByPolicy(result *v1alpha2.PolicyReportResult, policy policiesv1.Policy) (bool, error) {
+	policyName, err := getPolicyName(policy)
+	if err != nil {
+		return false, err
+	}
+	policyResourceVersion, hasPolicyResourceVersion := result.Properties[PropertyPolicyResourceVersion]
+	policyUID, hasPolicyUID := result.Properties[PropertyPolicyUID]
+	return result.Policy == policyName &&
+		hasPolicyResourceVersion && policyResourceVersion == policy.GetResourceVersion() &&
+		hasPolicyUID && types.UID(policyUID) == policy.GetUID(), nil
+}
+
+func findReusableResult(results []*v1alpha2.PolicyReportResult, policy policiesv1.Policy, resource unstructured.Unstructured) *v1alpha2.PolicyReportResult {
```

Comment:
I would document that for finding the result, we must ensure that:
policy has the same UID and ResourceVersion
resource has the same UID and ResourceVersion
To protect us from incorrectly relaxing this comparison in a future refactor.